### ISSUE TITLE
Feature request: cosine annealing scheduler with warm restarts

### ISSUE THREAD

**Issue by epwalsh** [allenai/allennlp#1642]
Currently, the pytorch implementation of [the cosine annealing learning rate schedule](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.CosineAnnealingLR) does not actually include warm restarts, which was the key idea in the paper referenced from the docs:

> It has been proposed in [SGDR: Stochastic Gradient Descent with Warm Restarts](https://arxiv.org/abs/1608.03983). Note that this only implements the cosine annealing part of SGDR, and not the restarts.

There is currently a [pull request open on pytorch](https://github.com/pytorch/pytorch/pull/6130) that addresses this issue, but it is unclear if or when it will be merged and released. Since the logic is relatively simple, and there are lots of potential variants of the algorithm, I propose implementing a custom cosine annealing schedule with warm restarts within AllenNLP.

I've actually already implemented this for my own research, so I'd be willing to submit a PR. It looks something like this:

```python
@LearningRateScheduler.register("cosine_with_restarts")
class CosineWithRestarts(_LRScheduler):
    """
    Cosine annealing with restarts.

    Parameters
    ----------
    optimizer : torch.optim.Optimizer

    T_max : int
        The maximum number of iterations within the first cycle.

    eta_min : float, optional (default: 0)
        The minimum learning rate.

    last_epoch : int, optional (default: -1)
        The index of the last epoch.

    """

    def __init__(self,
                 optimizer: torch.optim.Optimizer,
                 T_max: int,
                 eta_min: float = 0.,
                 last_epoch: int = -1,
                 factor: float = 1.) -> None:
        # pylint: disable=invalid-name
        self.T_max = T_max
        self.eta_min = eta_min
        self.factor = factor
        self._last_restart: int = 0
        self._cycle_counter: int = 0
        self._cycle_factor: float = 1.
        self._updated_cycle_len: int = T_max
        self._initialized: bool = False
        super(CosineWithRestarts, self).__init__(optimizer, last_epoch)

    def get_lr(self):
        """Get updated learning rate."""
        # HACK: We need to check if this is the first time get_lr() was called, since
        # we want to start with step = 0, but _LRScheduler calls get_lr with
        # last_epoch + 1 when initialized.
        if not self._initialized:
            self._initialized = True
            return self.base_lrs

        step = self.last_epoch + 1
        self._cycle_counter = step - self._last_restart

        lrs = [
            (
                self.eta_min + ((lr - self.eta_min) / 2) *
                (
                    np.cos(
                        np.pi *
                        ((self._cycle_counter) % self._updated_cycle_len) /
                        self._updated_cycle_len
                    ) + 1
                )
            ) for lr in self.base_lrs
        ]

        if self._cycle_counter % self._updated_cycle_len == 0:
            # Adjust the cycle length.
            self._cycle_factor *= self.factor
            self._cycle_counter = 0
            self._updated_cycle_len = int(self._cycle_factor * self.T_max)
            self._last_restart = step

        return lrs
```

**Comment by matt-gardner** [comment]
A PR for this seems reasonable to me.  You're right that the pytorch PRs seem pretty far from getting merged.  My only question is whether your code solves the issues that were brought up in the discussion of those pytorch PRs (I haven't yet looked in detail at the issues there or your code).
